src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -70,13 +70,28 @@ bool VMOperationTimeoutTask::is_armed() {
   return Atomic::load_acquire(&_armed) != 0;
 }
 
-void VMOperationTimeoutTask::arm() {
+void VMOperationTimeoutTask::arm(const char* vm_op_name) {
+  _vm_op_name = vm_op_name;
   _arm_time = os::javaTimeNanos();
   Atomic::release_store_fence(&_armed, 1);
 }
 
 void VMOperationTimeoutTask::disarm() {
   Atomic::release_store_fence(&_armed, 0);
+
+  // The two stores to `_armed` are counted in VM-op, but they should be
+  // insignificant compared to the actual VM-op duration.
+  jlong vm_op_duration = nanos_to_millis(os::javaTimeNanos() - _arm_time);
```

Comment:
You could move this ahead of the store. And then you only need a small comment in arm().

Reply:
You previously commented on slack: "Also disarm before getting the end time just to shorten the race with the watcher thread". I agree with that; for non-stuck VM-ops, checking timeout on VM thread is more reliable.

Reply:
Yep sorry forgot about that aspect.